### Micrograd uses scalars instead of vectors 
- most libraries use vectors(tensors)
- not really any difference other than complexity

In [1]:
from micrograd.engine import Value

## Value
- a variable type that creates a tree of all operations applied to the variable

In [5]:
a = Value(-4.0)
b = Value(2.0)

c = a + b
d = a * b + b**3

c += c + 1
c += 1 + c + (-a)

d += d * 2 + (b + a).relu() 
d += 3 * d + (b - a).relu() 

e = c - d
f = e**2
g = f / 2.0
g+= 10.0 / f

print(f'{g.data:.4f}')

24.7041


In this case, the a and b values will track all operations that all applied to them in a tree where a & b are respictevly the leaf nodes
 - this is considered the forward pass

In [6]:
g.backward()

- this method will work backword from the g node and recursively apply the chain rule, with respect to all internal nodes and input
- allows calculation of derivative with respect to a particular variable

In [7]:
print(f'{a.grad:.4f}')
print(f'{b.grad:.4f}')

138.8338
645.5773


- the partial derivatives aka dg/da and dg/db respecitvely
    - how are a and b effecting the output of g
- changing a positively will change g positively with a slope of 138.8338